### PR TITLE
 [CUDA Shared memory] Client side support

### PR DESCRIPTION
**Pull request by CoderHam**
- [x] Cuda shared memory client side support
- [x] Python cuda shared memory library (and example client)
- [x] fix server side changes to resolve bugs form server side implementation
- [x] Fix / revisit build checks for CUDA support in client container

PS: Require moving from Ubuntu to CUDA base image for client container

### CODE REVIEW

File: src/clients/c++/library/CMakeLists.txt [hunk 1]
```diff
@@ -86,9 +89,12 @@ add_library(
   TRTIS::request_static ALIAS request_static
 )
 if(${TRTIS_ENABLE_GPU})
-  target_include_directories(
+  target_include_directories(request_static PRIVATE ${CUDA_INCLUDE_DIRS})
+  target_link_libraries(
     request_static
-    PRIVATE ${CUDA_INCLUDE_DIRS}
+    PRIVATE -L/usr/local/cuda/lib64/stubs
```
Comment (deadeyegoodwin) [inline src/clients/c++/library/CMakeLists.txt]: Do we need stubs and nvidia-ml (I believe that is only needed for nvml metrics), or is just CUDA_LIBRARIES sufficient?

File: src/clients/c++/examples/CMakeLists.txt [hunk 2]
```diff
@@ -63,6 +63,9 @@ install(
 # ensemble_image_client
 #
 add_executable(ensemble_image_client ensemble_image_client.cc)
+if(${TRTIS_ENABLE_GPU})
```
Comment (deadeyegoodwin) [inline src/clients/c++/examples/CMakeLists.txt]: You shouldn't need these. request library rule already adds the target include so it should be inherited.

File: src/clients/c++/library/request_http.cc [hunk 3]
```diff
@@ -604,13 +600,16 @@ class SharedMemoryControlHttpContextImpl : public SharedMemoryControlContext {
 #if TRTIS_ENABLE_GPU
   Error RegisterCudaSharedMemory(
       const std::string& name, const cudaIpcMemHandle_t& cuda_shm_handle,
-      size_t byte_size, int device_id) override;
+      size_t byte_size, int device_id = 0) override;
```
Comment (deadeyegoodwin) [inline src/clients/c++/library/request_http.cc]: Don't default device_id

File: src/clients/c++/perf_client/CMakeLists.txt [hunk 4]
```diff
@@ -50,6 +50,9 @@ set(
 
 add_executable(perf_client
   ${PERF_CLIENT_SRCS} ${PERF_CLIENT_HDRS})
+if(${TRTIS_ENABLE_GPU})
```
Comment (deadeyegoodwin) [inline src/clients/c++/perf_client/CMakeLists.txt]: Remove

File: src/clients/python/CMakeLists.txt [hunk 5]
```diff
@@ -30,6 +30,9 @@ cmake_minimum_required (VERSION 3.5)
 # libcrequest.so
 #
 add_library(crequest SHARED crequest.cc)
+if(${TRTIS_ENABLE_GPU})
```
Comment (deadeyegoodwin) [inline src/clients/python/CMakeLists.txt]: Remove

File: src/clients/python/CMakeLists.txt [hunk 6]
```diff
@@ -40,12 +43,29 @@ if(NOT WIN32)
 # libcshm.so
 #
 add_library(cshm SHARED shared_memory/shared_memory.cc)
+if(${TRTIS_ENABLE_GPU})
```
Comment (deadeyegoodwin) [inline src/clients/python/CMakeLists.txt]: Why is this needed?

File: src/clients/python/CMakeLists.txt [hunk 7]
```diff
@@ -40,12 +43,29 @@ if(NOT WIN32)
...
 
+if(${TRTIS_ENABLE_GPU})
+#
+# libccudashm.so
+#
+add_library(ccudashm SHARED cuda_shared_memory/cuda_shared_memory.cc)
+target_include_directories(ccudashm PRIVATE ${CUDA_INCLUDE_DIRS})
+target_link_libraries(
+  ccudashm
+  PUBLIC -L/usr/local/cuda/lib64/stubs
```
Comment (deadeyegoodwin) [inline src/clients/python/CMakeLists.txt]: Remove stubs and nvidia-ml

File: src/clients/python/CMakeLists.txt [hunk 8]
```diff
@@ -40,12 +43,29 @@ if(NOT WIN32)
...
   rt
 )
 endif()
 
+if(${TRTIS_ENABLE_GPU})
+#
+# libccudashm.so
+#
+add_library(ccudashm SHARED cuda_shared_memory/cuda_shared_memory.cc)
+target_include_directories(ccudashm PRIVATE ${CUDA_INCLUDE_DIRS})
```
Comment (deadeyegoodwin) [inline src/clients/python/CMakeLists.txt]: PRIVATE -> PUBLIC so that it is inherited (I think)

File: src/clients/c++/library/CMakeLists.txt [hunk 9]
```diff
@@ -69,6 +69,9 @@ add_library(
 target_include_directories(
   request-library PRIVATE
   $<TARGET_PROPERTY:CURL::libcurl,INTERFACE_INCLUDE_DIRECTORIES>)
+if(${TRTIS_ENABLE_GPU})
+  target_include_directories(request-library PRIVATE ${CUDA_INCLUDE_DIRS})
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/clients/c++/library/CMakeLists.txt]: I think this change needs to be made for the request_static since that is inherited by the clients.

File: src/clients/c++/library/CMakeLists.txt (hunk 1 above)
Comment (CoderHam) [reply src/clients/c++/library/CMakeLists.txt]: You are right. CUDA_LIBRARIES should be enough

File: src/clients/c++/CMakeLists.txt [hunk 10]
```diff
@@ -29,3 +29,7 @@ cmake_minimum_required (VERSION 3.5)
 add_subdirectory(library)
 add_subdirectory(examples)
 add_subdirectory(perf_client)
+
+if(${TRTIS_ENABLE_GPU})
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/clients/c++/CMakeLists.txt]: True

File: src/clients/python/CMakeLists.txt (hunk 6 above)
Comment (CoderHam) [reply src/clients/python/CMakeLists.txt]: The shared memory handle is now shared (single handle representation) between the two libraries. This means that when GPU is enabled the shared memory library will also need to include cuda_runtime to be able to compile successfully.

File: Dockerfile.client [hunk 11]
```diff
@@ -24,10 +24,11 @@
 # (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-# Default setting is building on ubuntu:18.04
+# Default setting is building on nvidia/cuda:10.1-devel-ubuntu18.04
 ARG UBUNTU_VERSION=18.04
```
Comment (deadeyegoodwin) [inline Dockerfile.client]: Remove this arg? UBUNTU_VERSION no longer used?

File: src/clients/python/CMakeLists.txt [hunk 12]
```diff
@@ -30,6 +30,13 @@ cmake_minimum_required (VERSION 3.5)
 # libcrequest.so
 #
 add_library(crequest SHARED crequest.cc)
+if(${TRTIS_ENABLE_GPU})
+target_include_directories(crequest PUBLIC ${CUDA_INCLUDE_DIRS})
```
Comment (deadeyegoodwin) [inline src/clients/python/CMakeLists.txt]: indent

File: src/clients/python/CMakeLists.txt [hunk 13]
```diff
@@ -55,6 +77,9 @@ configure_file(__init__.py __init__.py COPYONLY)
 if(NOT WIN32)
 configure_file(shared_memory/__init__.py shared_memory/__init__.py COPYONLY)
 endif()
+if(${TRTIS_ENABLE_GPU})
+configure_file(cuda_shared_memory/__init__.py cuda_shared_memory/__init__.py COPYONLY)
```
Comment (deadeyegoodwin) [inline src/clients/python/CMakeLists.txt]: indent

File: src/core/provider.h [hunk 14]
```diff
@@ -337,7 +336,7 @@ class InferResponseProvider {
   InferRequestHeader request_header_;
 
   // Map from output name to the InferRequestHeader output information
-  // for that output.
+  // for that output.	  // for that output and it's memory type.
```
Comment (deadeyegoodwin) [inline src/core/provider.h]: Fix

File: Dockerfile.client (hunk 11 above)
Comment (CoderHam) [reply Dockerfile.client]: Added to make the user choose between CUDA Ubuntu 18.04 and CUDA Ubuntu 16.04 images. Removing and modifying documentation
